Lectures Review

brymz opened the pull request:
As per #372.

Formatting guides
- `>` marks actions, including reminders, demos, and exercises
- `##` major heading, `###` minor heading
- `*` for unordered list
- 'Key Terms' are capitalized and single quoted.
- Check code format and function

brymz commented on materials/databases_intro.md:
```diff
@@ -5,75 +5,79 @@ title: Introduction to Databases
+    * Cloud-based processing for really large data
+    * Can handle spatially explicit queries (GIS)
+* Improve quality control of data entry using scripts and tests
+    * vs. type constraints and use of forms in Access, Filemaker, etc.
```
Not sure what you are trying to say here about 'Improve quality control of data entry'.  I think clarifying this item could help link or motivate Exercise 1 - 'Data Validation in Excel'.

ethanwhite replied:
This is basically what I meant:

```
* Improve quality control of data entry
    * Type constraints
    * Use of forms in Access, Filemaker, etc.
```

brymz commented on materials/sql_joins.md:
````diff
@@ -46,3 +46,5 @@ SELECT species, scientificname
 FROM Main
 JOIN Species WHERE species = new_code;
 ```
````
Each of these examples of `JOIN` can have a little explanation about what the code is doing or what kind of shortcut is being taken. I would expect to really only use / need the 'Multi-table join' and 'Multi-table join with abbreviations' examples.

ethanwhite replied:
Agreed. Also let's go ahead and drop the "No ON clause" example since that's *definitely* not worth teaching anyone.

brymz commented on materials/other_data_structures_R.md:
```diff
@@ -5,102 +5,109 @@ title: Other Data Structures
 
-> Remember to downloade [*E. coli* genome data]({{ site.baseurl }}/data/ecoli_coding.fasta)
+> Remember to download [Archaea genome data]({{ site.baseurl }}/data/archaea_dna.zip).
  
```
@ethanwhite: Do you have an `ecoli_coding.fasta` file? How do you feel about using a data file pulled from the exercise (`Loops-multiple-files-R.md`)?

brymz commented on materials/debugging-R.md:
```diff
@@ -5,56 +5,61 @@ title: Debugging
+> Remember to
+>
+> * send `debugging-example-unfixed.R` to students so they can follow along.
+> * set up a new `Git` project that includes `debugging-example-unfixed.R`.
```
@ethanwhite: Do you have a `debugging-example-unfixed.R` file? From the lecture material, I think the file is a modification of `decomposition-example.R`. If you don't have the file, I should be able to revise `decomposition-example.R` to line up with the lesson.

brymz replied:
Files added and linked.

brymz commented:
Lectures Review COMPLETE. The PR is ready for review.

ethanwhite commented on materials/data_structures_R.md:
```diff
@@ -5,52 +5,77 @@ title: Data Structures
+* Subsetting:
+    * `counts[sites == 'a']`
+        * `==` means "equal to" in most languages. 
+        * Not `=`. `=` is used to assign `function()` arguments.
```
It's really used fro assignment more generally, so I think just "for assignment" is better here.

brymz replied:
`=` *can* be used for assignment more generally, but `<-` is preferred for assignment by the style guide. That nuance is why I chose to specifically state `function()` arguments.

ethanwhite replied:
I understand, but it's still a general assignment operator and the current language implies that it's only use of for function arguments. It's also by far the standard assignment operator in other languages, which is the real reason for `==`.

ethanwhite commented on materials/data_structures_R.md:
```diff
@@ -5,52 +5,77 @@ title: Data Structures
+    * `surveys[1:2,1:3]`
 
-***Exercise 3***
+> Assign remaining exercises.
```
Here I would "Do Exercise 3", bring the class back together, and make sure everyone understood and discuss. So, in cases like this I think it's better to keep the explicit Do Exercise. This should be carried through if there are similar changes in other sets of lectures.

ethanwhite commented on materials/database_structure.md:
```diff
@@ -7,44 +7,67 @@ language: R
+  
+![No redundant information example]({{ site.baseurl }}/materials/database_struct_no_redundant_information.png)
+
+* Use multiple tables to avoid redundant information. Unique `RecordID` link 
```
"Use a unique `RecordID`..."

ethanwhite commented on materials/database_structure.md:
```diff
@@ -7,44 +7,67 @@ language: R
+
+![Cross-tab table restructure]({{ site.baseurl }}/materials/databases_crosstab_restructured.png)
+
+### 5. No redundant information
```
We should add a bullet to this section along the lines of "Makes it more difficult to update the data if information changes. E.g., if a species name changes we want to be able to change it in one place, not hundreds of places. This is both easier and less error prone."

brymz replied:
Added to line 57

ethanwhite commented on materials/databases_intro.md:
```diff
@@ -5,81 +5,104 @@ title: Introduction to Databases
+* Simple and effective software for storing, managing and retrieving information
+* Widely used to store and maintain lots of existing data
+* Fast processing across large amounts of data
+    * Cloud-based processing for really large data
```
I'm not sure what you mean by "cloud-base" here (on a remote server?), and I'm not sure it will appeal to most novices. Maybe "Out-of-memory and remote processing for..."?

ethanwhite commented on materials/databases_intro.md:
```diff
@@ -5,81 +5,104 @@ title: Introduction to Databases
 
-## Relational databases
-
-**Demo in SQLite**
```
I would like to keep something like this note in. How about,

```
### Relationship databases

> Show students these properties using the Portal Project Teaching Database as an example

with Portal Project Teaching Database linked to the SQLite version of the DB.

ethanwhite commented on materials/databases_intro.md:
```diff
@@ -5,81 +5,104 @@ title: Introduction to Databases
     * A single observation or data point
 * Each column contains a single attribute
-    * A single type of information
+    * A single field or type of information
```
How about

```
* Each column (or field) contains a single attribute
    * A single type of information
```

ethanwhite commented on materials/databases_intro.md:
```diff
@@ -5,81 +5,104 @@ title: Introduction to Databases
   records/row
-* This is called the primary key
-* The easiest way to do this is to use an Integer that increments every time a
+* This is called the 'Primary Key'
```
Are these supposed to be backticks? Same question for next line.

brymz replied:
I am trying to distinguish 'Key Terms' from code `objects`, but sometimes that distinction is a bit ambiguous. Is 'Primary Key' a key term while `Integer` is an object?

ethanwhite replied:
OK. I'm fine with it this way, just hadn't seen a lot of `'stuff'` so thought I'd check.

ethanwhite commented on materials/databases_intro.md:
```diff
@@ -5,81 +5,104 @@ title: Introduction to Databases
+
+1\. Enter the data into another program (e.g., a spreadsheet) and import it to SQLite
+
+> Do [Exercise 1 - Data entry validation in Excel]({{ site.baseurl }}/exercises/Qaqc-data-entry-validation-in-excel-SQL/). 
```
This exercise should be associated with an introduction to data validation in Excel. Were there not any lecture notes on this?

brymz replied:
There were no notes on this in `materials/databases_intro.md` or `lectures/sql_intro.md`. There is a `materials/access-forms.md` that might include relevant info, but nothing specifically to Excel.

ethanwhite replied:
Hmmm... I thought I had something, but I can't find it either. Can you develop something from the reading when you get the chance?

ethanwhite commented on materials/databases_intro.md:
```diff
@@ -5,81 +5,104 @@ title: Introduction to Databases
+> Demo [Updating Tables]({{ site.baseurl }}/materials/sql-updating-tables).
+>  
+> Do [Exercise 4 - Adding Records]({{ site.baseurl }}/exercises/Database-control-adding-records-SQL/) and [Exercise 5 - Updating Records]({{ site.baseurl }}/exercises/Database-control-updating-records-SQL/).
+
```
In general I think this section should start with a quick overview of the options, not offset with exercises, and then provide brief introductions to how to do the thing in the bullet point before having the students work on the exercise. It looks like some of the introductory material is in separate files, but it seems like it would be most useful to integrate it into the lecture directly. In cases where that material doesn't exist it will need to be added, either now or by creating new issues and taking care of it later.

ethanwhite commented on materials/debugging-R.md:
```diff
@@ -5,56 +5,63 @@ title: Debugging
+
+* Run the code line by line checking each step.
+* More difficult once functions are involved.
+    * Isolate functions.
```
I'm not sure what you mean here. Can you clarify?

ethanwhite commented on materials/debugging-R.md:
```diff
@@ -5,56 +5,63 @@ title: Debugging
+* Run the code line by line checking each step.
+* More difficult once functions are involved.
+    * Isolate functions.
+    * Use `print()`.
```
Not sure we need this here since it then occurs immediately below.

ethanwhite commented on materials/debugging-R.md:
```diff
@@ -5,56 +5,63 @@ title: Debugging
 ## Debugger
 
-* Like a print statement but better
+* Advanced error visualization (`print()`) technique
```
I like the original version better here. I'm open to a revision but without knowing what was there before I wouldn't understand what this was meant to say. For lecture I think that simple intros like "like X, but better" can be fine.

ethanwhite commented on materials/debugging-R.md:
```diff
@@ -64,20 +71,25 @@ language: R
-4. Environment gives the current values for all variables (Local
-   vs. Global). Can switch scopes with dropdown.
-
+3. Program runs until breakpoint and stops ( *before executing the line of code* )
```
extra spaces

ethanwhite commented on materials/dplyr.md:
```diff
@@ -5,79 +5,102 @@ title: Working with Tabular Data
+> 
+> * download [`surveys.csv`](https://ndownloader.figshare.com/files/2292172).   
+> * download [`portal_mammals.sqlite`](https://ndownloader.figshare.com/files/2292171).
+> * `setwd()` in R.
```
We're teaching projects and relative paths, not `setwd()`... or maybe I'm misunderstanding something here.

brymz replied:
It's a reminder to the instructor to set the working directory in their R console before the start of class to be able to access `surveys.csv` without a path, such as `read.csv("surveys.csv")`.

ethanwhite replied:
Right, but they shouldn't be doing that they should be putting `surveys.csv` in their working directory by using an RStudio Project.

brymz replied:
Ok, that makes sense to me. This is a good time to point out that I don't use RStudio Projects or Git from RStudio and did not test those aspects of the `lectures/`. In these cases, my review was more conceptual.

ethanwhite commented on materials/dplyr.md:
```diff
@@ -5,79 +5,102 @@ title: Working with Tabular Data
+
+* Main way that reusable code is shared in R
+* Combination of code, data, and documentation
+* Download and instal packages with the R console:
```
install

ethanwhite commented on materials/dplyr.md:
```diff
@@ -5,79 +5,102 @@ title: Working with Tabular Data
+        * Can result in namespace issues when different packages have functions with the same name 
+        * `filter()` is the name of a function in `dplyr` and `stats` packages
+    * Call a function in the package directly: `::`
+        * No import needed
```
"No use of `library` needed"?

ethanwhite commented on materials/dplyr.md:
```diff
@@ -5,79 +5,102 @@ title: Working with Tabular Data
+        * `%>%`
+    * Operation: 
+        * `x %>% f(y)` translates to `f(x, y)`
+        * `%>%` takes the output of one command and passes it as input to the next command 
```
This should be the first bullet since it's a clear plain language description

ethanwhite commented on materials/dplyr.md:
```diff
@@ -5,79 +5,102 @@ title: Working with Tabular Data
+  filter(species_id == "DS") %>%
   group_by(year) %>%
-  summarize(avg_weight = mean(weight))
+  summarize(avg_weight = mean(weight, na.rm=T))
```
`TRUE` not `T` since `TRUE` is best practice.

ethanwhite commented on materials/for_loops_R.md:
```diff
@@ -63,16 +79,29 @@ biomass_data <- data.frame(exper1 = c(24, 32, 62),
+    * Other versions allow you various control options
+        * [`lapply()`](http://finzi.psych.upenn.edu/R/library/base/html/lapply.html): Operate across lists and vectors
+        * [`mapply()`](http://finzi.psych.upenn.edu/R/library/base/html/mapply.html): Pass multiple variables or function arguments
+        * [`rapply()`](http://finzi.psych.upenn.edu/R/library/base/html/rapply.html): Recursive operation
```
I think getting into recursion is out of scope. Let's add `sapply` here instead since that's often what folks really want.

ethanwhite commented on materials/for_loops_R.md:
```diff
@@ -63,16 +79,29 @@ biomass_data <- data.frame(exper1 = c(24, 32, 62),
+        * [`mapply()`](http://finzi.psych.upenn.edu/R/library/base/html/mapply.html): Pass multiple variables or function arguments
+        * [`rapply()`](http://finzi.psych.upenn.edu/R/library/base/html/rapply.html): Recursive operation
+    * Why use `apply()`:
+        * Speed 
```
The whole speed thing is generally incorrect (see: https://stackoverflow.com/questions/2275896/is-rs-apply-family-more-than-syntactic-sugar), which is what I often say here, so we should probably be a little clearer about that. The main argument for their use of better readability, which I think can be true up to a point, but not if things get to complex, which is basically captured in the second point, but could probably be rephrased a little so this is clear to folks seeing this material for the first time.

ethanwhite commented on materials/functions_R.md:
```diff
@@ -0,0 +1,95 @@
+### Understandable chunks
+
+* Human brain can only hold ~7 things in memory.
+    * Write programs with no more than ~7 things.
```
"Write programs..." -> "Write programs that don't require remembering more than ~7 things at once."

ethanwhite commented on materials/functions_R.md:
```diff
@@ -0,0 +1,95 @@
+
+> Assign Exercises 
+>
+> * [1 - Use and Modify]({{ site.baseurl }}/exercises/Functions-use-and-modify-R)
```
Let's move up Exercise 1 to before "Treat functions like a black box", but just assign parts 1 & 2 of the exercise. This will give students a chance to fail early and then ask for clarification + get an initial exposure, try it, and then see another example.

ethanwhite commented on materials/functions_R.md:
```diff
@@ -0,0 +1,95 @@
+> Assign Exercises 
+>
+> * [1 - Use and Modify]({{ site.baseurl }}/exercises/Functions-use-and-modify-R)
+> * [2 - Writing Functions]({{ site.baseurl }}/exercises/Functions-writing-functions-R)
```
The here, let's have them do 1.3 and 2.

ethanwhite commented on materials/functions_R.md:
```diff
@@ -0,0 +1,95 @@
+>
+> * [1 - Use and Modify]({{ site.baseurl }}/exercises/Functions-use-and-modify-R)
+> * [2 - Writing Functions]({{ site.baseurl }}/exercises/Functions-writing-functions-R)
+> * [3 - Nested Functions]({{ site.baseurl }}/exercises/Functions-nested-functions-R)
```
Let's add a brief chunk to the lecture before this line introducing the idea of nesting. Experience last year suggested this was too big of a jump without some guidance for about half the class..

ethanwhite commented on materials/r-intro.md:
```diff
@@ -5,47 +5,56 @@ title: Introduction to R
 * Start a new project
 * Create a `data` subdirectory
-* Add an example csv file
+* Add an example `.CSV` file
```
lower case for `.csv`

brymz replied:
I have used capitalized file extensions throughout the course as a 'standard' format. Do you want all of the them changed?

ethanwhite replied:
Oh, sorry I hadn't noticed this before. Yes, we should definitely change them to lower case since almost every use of them in the real world is lower case and most operating systems and programming languages are case sensitive.

brymz replied:
Ok. Should I use backticks (`csv`)? Do I include the `.` (`.csv`)?

ethanwhite replied:
I think `csv` is fine, but I don't have any strong feelings about the dot, so whatever you think is best.

ethanwhite commented on materials/r-intro.md:
```diff
@@ -5,47 +5,56 @@ title: Introduction to R
-  (possibly with subdirectories of their own)
-* src in a separate subdirectory
+* code in top-level directory with subdirectories for data, results, etc
+  ( *possibly with subdirectories of their own* )
```
remove extra spaces

ethanwhite commented on materials/r-sql.md:
```diff
@@ -5,11 +5,17 @@ title: Integrating R and SQL
-* Use `DBI` and `RSQLite` (or other appropriate DBMS plugin)
+> Remember to
+>    
+> * download [`portal_mammals.sqlite`](https://ndownloader.figshare.com/files/2292171). *Make sure it does not have `SpeciesCounts` view.*
```
That file doesn't look like it has that view, so do we need the italicized text?

brymz replied:
I added this note after an experience I had testing the lecture. When I ran the code a second time, I got an error that the `SpeciesCounts` already existed. Could other instructors run into this same issue? Do they need a reminder to get a fresh version?

ethanwhite replied:
Oh yeah. I can see that happening after a practice run. How about "*Make sure the copy you are going to use in class does not have..."

ethanwhite commented on materials/r-sql.md:
```diff
@@ -5,11 +5,17 @@ title: Integrating R and SQL
+> Remember to
+>    
+> * download [`portal_mammals.sqlite`](https://ndownloader.figshare.com/files/2292171). *Make sure it does not have `SpeciesCounts` view.*
+> * `setwd()` in R.
```
Drop `setwd()`

ethanwhite commented on materials/r-sql.md:
````diff
@@ -18,42 +24,46 @@ portalDB <- "portal_mammals.sqlite"
+dbListTables(conn)
 ```
 
-* Show that this table now exists in the database
````
I think it would be good to keep this using the new `> ` format in the same way that you did for showing them the view. I like showing this immediately before doing running and showing the view change.

brymz replied:
Added before and after actions.

ethanwhite commented on materials/sql_nested_queries.md:
```diff
@@ -5,25 +5,29 @@ title: Nested Queries
+### Take Home Message
+
+* Basically, any object or value following a query command statement in SQL 
+( *including tables, conditions, and even values* ) can be replaced with another 
```
extra spaces

ethanwhite commented on materials/version-control-R.md:
```diff
@@ -0,0 +1,209 @@
+language: R
+--- 
+
+> Remind students to setup a GitHub account and email me their username.
```
"email me" -> "email the instructor"

ethanwhite commented on materials/version-control-R.md:
```diff
@@ -0,0 +1,209 @@
+
+> Remind students to setup a GitHub account and email me their username.
+
+> Arrange for partner to `push` following code for 'Collaborating' demo.
```
"Arrange to have a teaching partner attend class and `push` the following code for the 'Collaborating' demo."

ethanwhite commented on materials/version-control-R.md:
```diff
@@ -0,0 +1,209 @@
+
+* Who has a directory on their computer with a bunch of filenames
+    * [Like this?](http://www.phdcomics.com/comics/archive.php?comicid=1531)
+    * [Or like this?](http://www.phdcomics.com/comics/archive.php?comicid=1323)
```
Add note in the prep section to open these pages in a browser and zoom in to make the images fill the screen.

brymz replied:
Done.

ethanwhite commented on materials/version-control-R.md:
```diff
@@ -0,0 +1,209 @@
+
+### Benefits of version control
+
+* Track changes ( *but better* )
```
Extra spaces.

ethanwhite commented on materials/version-control-R.md:
```diff
@@ -0,0 +1,209 @@
+
+* Track changes ( *but better* )
+    * Tracks every change ever made in groups called commits
+	* Every commit stores the full state of all of your files at that time
```
These red marks are probably indicating that I originally used of tabs instead of spaces. Let's go ahead and replace them with spaces.

ethanwhite commented:
*Really* nice job @brymz and a ton of work. Thanks so much. As you've already seen I've made minor comments throughout. Keeping sending along the questions and let me know when you've made all the tweaks and we'll get this in.

brymz commented:
Revisions as per @ethanwhite comments is COMPLETE. Ready for merge pending follow-up review.

ethanwhite commented on materials/database-control-sql.md:
```diff
@@ -0,0 +1,153 @@
+> Do [Exercise 2 - Importing Data]({{ site.baseurl }}/exercises/Database-control-importing-data-SQL), Task 11.
+
+### Enter the data directly
+
```
In the past what I've done for this section is just show them how to add a record to an existing table using the SQLite manager GUI (the `Add` button on the `Browse and Search` tab). I think getting into  the details of how to create a new table and add data using SQL is too much for the first week. This is a nice description though so let's put it in a separate lecture that could be used later in the course or by folks wanting to teach more SQL.

ethanwhite replied:
Arg, looks like I'm probably asking you to do something that I asked you to do in the first place. Sorry about that. I do think we need to pull the detailed treatment of creating tables and adding data using SQL though.

brymz replied:
No worries. Just part of the process. I'll probably have to modify the exercise, too, if I understand you correctly.

brymz replied:
> I'll probably have to modify the exercise, too.

Not true. Looks like the exercise is general enough to be solved both ways.

brymz replied:
I have modified `database-control-sql.md` to a SQLite Manager workflow and moved the workflow using queries to `databse-control-queries-sql.md`. 

I noticed now that there is no introduction to basic queries `WHERE`, `DISTINCT`, `ORDER BY`, etc.

ethanwhite replied:
I can cleanup any associated issues when I prep this section.

ethanwhite commented on materials/sql_joins.md:
```diff
@@ -7,44 +7,65 @@ language: SQL
+    * based on condition
+* This query selects `species` records in the `PortalMammals_main` table and 
+  `scientificname` found in the `PortalMammals_species` table.
+    * The query links the `species` with `new_code` from `PortalMammals_species`.
```
Some of these names don't match the current values in the Portal Project Teaching Database, but I've added a separate issue for that: #441.

ethanwhite commented on materials/sql_joins.md:
```diff
@@ -7,44 +7,65 @@ language: SQL
-JOIN Species ON species = new_code;
+SELECT species, scientificname
+FROM PortalMammals_main
+JOIN PortalMammsls_species WHERE species = new_code;
```
Why are we introducing `WHERE` here instead of just using `ON` from the start?

brymz replied:
The examples are a selection from a much longer list of JOIN examples that was in the 'original' version. I liked/kept the `WHERE` example because I liked the conceptual step of `WHERE` being functionally similar to the `ON`. I don't mind going with `ON` from the start.

ethanwhite replied:
Yeah, I waffle back and forth on that every year. The upside is as you say, the downside is that demoing the wrong way to do things tends to confuse beginners (as I now regularly notice at my daughter's martial art's class). Let's leave it as is for the moment and if I think it needs changing when I'm reviewing the material for class I'll make the changes myself.

brymz replied:
I wonder what the evidence-based pedagogy says?

ethanwhite replied:
I recall reading somewhere not to show the wrong way or at least not to show it first, but I don't remember where so I'm not sure if my brain is just inventing that recollection.

brymz replied:
'Fortunately' then, using `WHERE` is only semantically wrong and doesn't break anything.
 
@gvwilson recently shared a teaching philosophy with me that suggests that sometimes the best learning strategy is to first show a bad example and then make the fixes to demonstrate the best practice. Maybe it matters where where the students are on the novice-expert shift.

ethanwhite commented:
Looks awesome. A couple of new suggestions/questions and then we'll get this in.

brymz commented:
Updated SQL lectures and removed out-dated `portal_mammals.sqlite` from `data/`. This is ready to go! I will check `exercises/` and `solutions/` as per #441.

ethanwhite commented:
Thanks @brymz! Great work. In it goes.

brymz commented:
That was fun! I had a great place to start.
